README.md
# NAME 

re::engine::Hyperscan - High-performance regular expression matching library (Intel only)

# SYNOPSIS

    use re::engine::Hyperscan;

    if ("Hello, world" =~ /(Hello|Hi), (world)/) {
        print "Greetings, $1!";
    }

# DESCRIPTION

ALPHA - Does not work YET

Replaces perl's regex engine in a given lexical scope with Intel's 
Hyperscan regular expressions provided by `libhyperscan`.

This provides the fastest regular expression library on Intel-CPU's
only, but needs to fall back to the core perl regexp compiler with
backtracking, lookbehind, zero-width assertions and more advanced
patterns.  It is typically 50% faster then the core regex engine.

For the supported syntax see
[https://01org.github.io/hyperscan/dev-reference/compilation.html](https://01org.github.io/hyperscan/dev-reference/compilation.html).

With the following unsupported constructs in the pattern, the compiler
will fall back to the core re engine:

- Backreferences and capturing sub-expressions.
- Arbitrary zero-width assertions.
- Subroutine references and recursive patterns.
- Conditional patterns.
- Backtracking control verbs.
- The `\C` "single-byte" directive (which breaks UTF-8 sequences).
- The `\R` newline match.
- The `\K` start of match reset directive.
- Callouts and embedded code.
- Atomic grouping and possessive quantifiers.

# METHODS

- min\_width (RX)

    Returns the result from hs\_expression\_info(). NYI
    The minimum length in bytes of a match for the pattern.

- max\_width (RX)

    Returns the result from hs\_expression\_info(). NYI
    The maximum length in bytes of a match for the pattern. If the pattern
    has an unbounded maximum width, this will be set to the maximum value of
    an unsigned int (UINT\_MAX).

- matches\_at\_eod (RX)

    Returns the result from hs\_expression\_info(). NYI
    Whether this expression can produce matches at end of data (EOD).

- matches\_only\_at\_eod (RX)

    Returns the result from hs\_expression\_info(). NYI
    Whether this expression can \*only\* produce matches at end of data (EOD).

# FUNCTIONS

- ENGINE

    Returns a pointer to the internal Hyperscan engine, the database,
    suitable for the XS API `(regexp*)re->engine` field.

# AUTHORS

Reini Urban <rurban@cpan.org>

# COPYRIGHT

Copyright 2017 Reini Urban.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for re::engine::Hyperscan

0.02

    - major fixes to enable the compiler
    - add META_MERGE resources, esp. for the bugtracker

0.01 Fri Apr 7 19:27:31 2017 +0200 rurban

    - Started from re-engine-PCRE2.
      Fails only a few tests.


Makefile.PL
use 5.010;
use strict;
use warnings;
use ExtUtils::MakeMaker 6.31;
use Config;

my $prefix;
if (@ARGV) { # PREFIX=~
  while (my $arg = shift @ARGV) {
    if ($arg =~ /^-?-?PREFIX=(.*)/i) {
      # lib and include prefix if installed locally (ie. travis)
      $prefix = $1;
    }
  }
}

eval 'use Devel::CheckLib;' and
  check_lib_or_exit(
    lib => 'hs',
    header => 'hs/hs.h',
    $prefix ? (
      incpath => "$prefix/include",
      libpath => "$prefix/lib"
    ) : ());

my %WriteMakefileArgs = (
  'NAME'         => 're::engine::Hyperscan',
  'VERSION_FROM' => 'Hyperscan.pm',
  'ABSTRACT'     => 'Fast Hyperscan regular expression engine - Intel-only',
  'AUTHOR'       => 'Reini Urban <rurban@cpan.org>',
  'BUILD_REQUIRES' => {
    'Data::Dumper' => '0',
    'Test::More'   => '0.88',
    ($ENV{RELEASE_TESTING}
     ? ('Test::Kwalitee'    => 0,
        'Class::XSAccessor' => 0,
        'Text::CSV_XS'      => 0,
        'List::MoreUtils'   => 0,
        'Pod::Spell::CommonMistakes' => 0,
        'Test::Pod'           => '1.00',
        'Test::Pod::Coverage' => '1.04',
        'Test::Spelling'      => 0)
     : ()),
  },
  'CONFIGURE_REQUIRES' => {
    'Devel::CheckLib'     => '1.00',
    'ExtUtils::MakeMaker' => '6.31'
  },
  'dist' => {
    PREOP	=> 'pod2markdown $(VERSION_FROM) | tee README.md >$(DISTVNAME)/README.md;'
                  .' chmod -R u=rwX,go=rX . ;',
    COMPRESS	=> 'gzip -9v',
    SUFFIX	=> '.gz',
  },
  'EXE_FILES' => [],
  ($prefix
   ? ('CCFLAGS'   => "$Config{ccflags} -I$prefix/include",
      'LDDLFLAGS' => "$Config{lddlflags} -I$prefix/lib")
   : ()),
  'LIBS' => ($prefix ? [ "-L$prefix/lib", '-lhs' ] : [ '-lhs' ]),
  'LICENSE' => 'perl',
  'test' => {
    'TESTS' => 't/*.t t/perl/*.t'
  },
  'SIGN' => 1,
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

package MY;

sub top_targets {
  local $_ = shift->SUPER::top_targets(@_);
  s/\$\(FIRST_MAKEFILE\) blibdirs/\$(FIRST_MAKEFILE\) blibdirs README.md/;
  $_
}

sub depend {
  "
README : \$(VERSION_FROM)
	pod2text \$(VERSION_FROM) > README

README.md : \$(VERSION_FROM)
	pod2markdown \$(VERSION_FROM) > \$@

release : dist
	echo git commit -a -m\"release \$(VERSION)\"
	git tag \$(VERSION)
	cpan-upload \$(DISTVNAME).tar\$(SUFFIX)
	git push
	git push --tags

test_cover :: pure_all
	\$(RM_RF) cover_db
	\$(PERLRUNINST) -S cover -test
test_coveralls :: pure_all
	\$(PERLRUNINST) -S cover -test -report coveralls

gcov : \$(BASEEXT).c.gcov \$(BASEEXT).gcov cover_db/\$(BASEEXT)-xs.html

\$(BASEEXT).c.gcov \$(BASEEXT).xs.gcov : \$(BASEEXT).xs
	\$(MAKE) CCFLAGS=\"\$(CCFLAGS) -fprofile-arcs -ftest-coverage\" LDDLFLAGS=\"\$(LDDLFLAGS) -fprofile-arcs -ftest-coverage\"
	gcov \$(BASEEXT).c \$(BASEEXT).xs

cover_db/\$(BASEEXT)-xs.html : \$(BASEEXT).xs.gcov
	PERL5OPT=-MDevel::Cover make test
	-$^X -S gcov2perl \$(BASEEXT).c.gcov \$(BASEEXT).xs.gcov
	$^X -S cover

gprof :
	\$(MAKE) CCFLAGS=\"\$(CCFLAGS) -pg\" LDDLFLAGS=\"\$(LDDLFLAGS) -pg\"
"
}


Hyperscan.pm
package re::engine::Hyperscan;
our ($VERSION, $XS_VERSION);
BEGIN {
  $VERSION = '0.02';
  $XS_VERSION = $VERSION;
  $VERSION = eval $VERSION;
}
use 5.010;
use strict;
use XSLoader ();

# All engines should subclass the core Regexp package
our @ISA = 'Regexp';

BEGIN {
  XSLoader::load;
}

sub import {
  $^H{regcomp} = ENGINE;
}

sub unimport {
  delete $^H{regcomp} if $^H{regcomp} == ENGINE;
}

1;

__END__

=head1 NAME 

re::engine::Hyperscan - High-performance regular expression matching library (Intel only)

=head1 SYNOPSIS

    use re::engine::Hyperscan;

    if ("Hello, world" =~ /(Hello|Hi), (world)/) {
        print "Greetings, $1!";
    }

=head1 DESCRIPTION

ALPHA - Does not work YET

Replaces perl's regex engine in a given lexical scope with Intel's 
Hyperscan regular expressions provided by F<libhyperscan>.

This provides the fastest regular expression library on Intel-CPU's
only, but needs to fall back to the core perl regexp compiler with
backtracking, lookbehind, zero-width assertions and more advanced
patterns.  It is typically 50% faster then the core regex engine.

For the supported syntax see
L<https://01org.github.io/hyperscan/dev-reference/compilation.html>.

With the following unsupported constructs in the pattern, the compiler
will fall back to the core re engine:

=over

=item Backreferences and capturing sub-expressions.

=item Arbitrary zero-width assertions.

=item Subroutine references and recursive patterns.

=item Conditional patterns.

=item Backtracking control verbs.

=item The C<\C> "single-byte" directive (which breaks UTF-8 sequences).

=item The C<\R> newline match.

=item The C<\K> start of match reset directive.

=item Callouts and embedded code.

=item Atomic grouping and possessive quantifiers.

=back

=head1 METHODS

=over

=item min_width (RX)

Returns the result from hs_expression_info(). NYI
The minimum length in bytes of a match for the pattern.

=item max_width (RX)

Returns the result from hs_expression_info(). NYI
The maximum length in bytes of a match for the pattern. If the pattern
has an unbounded maximum width, this will be set to the maximum value of
an unsigned int (UINT_MAX).

=item unordered_matches (RX)

Returns the result from hs_expression_info(). NYI
Whether this expression can produce matches that are not returned in
order, such as those produced by assertions.

=item matches_at_eod (RX)

Returns the result from hs_expression_info(). NYI
Whether this expression can produce matches at end of data (EOD).

=item matches_only_at_eod (RX)

Returns the result from hs_expression_info(). NYI
Whether this expression can *only* produce matches at end of data (EOD).

=back

=head1 FUNCTIONS

=over

=item ENGINE

Returns a pointer to the internal Hyperscan engine, the database,
suitable for the XS API C<<< (regexp*)re->engine >>> field.

=back

=head1 AUTHORS

Reini Urban <rurban@cpan.org>

=head1 COPYRIGHT

Copyright 2017 Reini Urban.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
